beat/backend.py

```diff
@@ -491,6 +491,175 @@ def point(self, idx):
+        """
+        Write the lpoint to file. If data_to_write is None it will try to write from buffer.
+        :param data_to_write: A lpoint data, expected an list of numpy arrays.
+        """
+        # Write binnary
+        if data_to_write is None and self.buffer == []:
+            logger.info("There is no data to write into file.")
+            raise ValueError("There is no data to write into file.")
+
+        try:
+            # create initial data using the data structure.
+            data = num.zeros(1, dtype=self.__data_structure)
+
+            with open(self.filename, mode="ab+") as file:
+                if data_to_write is None:
+                    for lpoint, draw in self.buffer:
+                        for names, array in zip(self.varnames, lpoint):
+                            data[names] = array
+                        data.tofile(file)
+                else:
+                    for names, array in zip(self.varnames, data_to_write):
+                        data[names] = array
+                    data.tofile(file)
+        except EnvironmentError as e:
+            print("Error on write file: ", e)
+
+    def record_buffer(self):
+
+        n_samples = len(self.buffer)
+        self.stored_samples += n_samples
+
+        if not self.progressbar:
+            if n_samples > self.buffer_size // 2:
+                logger.info(
+                    'Writing %i / %i samples of chain %i to disk...' %
+                    (self.stored_samples, self.draws, self.chain))
+
+        t0 = time()
+        logger.debug(
+            'Start Record: Chain_%i' % self.chain)
+        self.__write_data_to_file()
+        # for lpoint, draw in self.buffer:
+        #    self.record(lpoint, draw)
+
+        t1 = time()
+        logger.debug('End Record: Chain_%i' % self.chain)
+        logger.debug('Writing to file took %f' % (t1 - t0))
+        self.empty_buffer()
+
+    def record(self, lpoint, draw):
+        """
+        Record results of a sampling iteration.
+
+        Parameters
+        ----------
+        lpoint : List of variable values
+            Values mapped to variable names
+        """
+
+        # here we maybe should include a method to check if lpoint has a right size.
+        # columns = itertools.chain.from_iterable(
+        #   map(str, value.ravel()) for value in lpoint)
+        # data = num.array([tuple(columns)], dtype=self.__data_structure)
+        data = num.zeros(1, dtype=self.__data_structure)
+        for names, array in zip(self.varnames, lpoint):
+            data[names] = array
+
+        logger.debug('Writing...: Chain_%i step_%i' % (self.chain, draw))
+
+        self.__write_data_to_file(lpoint)
+        # Write binnary
+        #try:
+        #    with open(self.filename, mode="ab+") as file:
+        #        data.tofile(file)
+        #except EnvironmentError as e:
+        #    print("Error on write file: ", e)
+
+    def _load_df(self):
```

Comment:
does it get variable names and shapes from the file header? this has to work without having the variable names as an input, this has to be somehow stored in the binary file or in an extra header file, this could be a single file which is valid for many Chains

Reply:
please check the function: backend.extract_variables_from_df
maybe we need something like this also for this backend ...